evdev/uinput.py

```diff
@@ -281,11 +283,77 @@ def _verify(self):
             msg = 'uinput device name must not be longer than {} characters'
             raise UInputError(msg.format(_uinput.maxnamelen))
 
-    def _find_device(self):
+    def _find_device(self, fd):
+        '''
+        Tries to find the device node. Will delegate this task to one of
+        several platform-specific functions.
+        '''
+        if platform.system() == 'Linux':
+            try:
+                sysname = _uinput.get_sysname(fd)
+                return self._find_device_linux(sysname)
+            except OSError:
+                # UI_GET_SYSNAME returned an error code. We're likely dealing with
+                # an old kernel. Guess the device based on the filesystem.
+                pass
+
+        # If we're not running or Linux or the above method fails for any reason,
+        # use the generic fallback method.
+        return self._find_device_fallback()
+
+    def _find_device_linux(self, sysname):
+        '''
+        Tries to find the device node when running on Linux.
+        '''
+
+        syspath = f'/sys/devices/virtual/input/{sysname}'
+
+        # The sysfs entry for event devices should contain exactly one folder
+        # whose name matches the format "event[0-9]+". It is then assumed that
+        # the device node in /dev/input uses the same name.
+        regex = re.compile('event[0-9]+')
+        for entry in os.listdir(syspath):
+            if regex.fullmatch(entry):
+                device_path = f'/dev/input/{entry}'
+                break
+        else:  # no break
+            raise FileNotFoundError()
+
+        # It is possible that there is some delay before /dev/input/event* shows
+        # up on old systems that do not use devtmpfs, so if the device cannot be
+        # found, wait for a short amount and then try again once.
+        try:
+            return device.InputDevice(device_path)
+        except FileNotFoundError:
+            time.sleep(0.1)
+            return device.InputDevice(device_path)
+
+    def _find_device_fallback(self):
+        '''
+        Tries to find the device node when UI_GET_SYSNAME is not available or
+        we're running on a system sufficiently exotic that we do not know how
+        to interpret its return value.
+        '''
         #:bug: the device node might not be immediately available
         time.sleep(0.1)
 
+        # Strictly speaking, we cannot be certain that everything returned by list_devices()
+        # ends at event[0-9]+: it might return something like "/dev/input/events_all". Find
+        # the devices that have the expected structure and extract their device number.
```

Comment:
```suggestion
        # There could also be another device with the same name already present,
        # make sure to select the newest one.
        # Strictly speaking, we cannot be certain that everything returned by list_devices()
        # ends at event[0-9]+: it might return something like "/dev/input/events_all". Find
        # the devices that have the expected structure and extract their device number.
```